### PR TITLE
Feature | OSS124 Invite User Flow

### PR DESCRIPTION
**Pull request by msinnes**
https://synapse.leankit.com/Boards/View/285322106/464218052

#49

### CODE REVIEW

File: api/src/application/client/client-service.js [hunk 1]
```diff
@@ -88,10 +88,15 @@ module.exports = (bookshelf) => ({
       .fetch({ withRelated: clientRelationships });
   },
 
+  findRedirectUriByClientId(clientId) {
```
Comment (spruce-bruce) [inline api/src/application/client/client-service.js]: Clients can have many redirect uris configured. You won't be able to make a decision about which redirect uri to use based solely on the clientId. You need to update the invite and resent-invite routes to expect a `redirect_uri` body parameter so that the app can tell OIDC _which_ redirect uri to use.

You must then validate that the redirect uri provided matches one of the redirect_uris configured for the client

File: api/migrations/20170509230917_add_email_id_column.js [hunk 2]
```diff
@@ -1,20 +1,8 @@
...
-        `);
-      } else {
-        users.forEach(user => {
-          const emailID = user.email.toLowerCase();
-          knex('SIP_user').update('email_id', emailID).where('id', user.id);
-        });
-
-      }
-  }));
+        return table.string('email_id').notNull();
```
Comment (spruce-bruce) [inline api/migrations/20170509230917_add_email_id_column.js]: What is the email_id field for? I see in this PR where you're doing a `WHERE` on this but I don't see any code in this PR that inserts or updates info into this column. Is this from some other PR and you're just simplifying the migration?

and if email_id is necessary, why are you removing all the code that sets `email_id` for existing users?

File: api/src/application/api/api-routes.js [hunk 3]
```diff
@@ -7,7 +7,11 @@ module.exports = (userService, mixedValidation, rowNotExists, rowExists) => [
     method: 'POST',
     path: '/api/invite',
     handler: (request, reply) => {
-      reply(userService.inviteUser(request.payload));
```
Comment (spruce-bruce) [inline api/src/application/api/api-routes.js]: This line should work exactly the same as what you replaced it with. Why the change?

File: api/src/application/user/user-service.js [hunk 4]
```diff
@@ -30,49 +31,73 @@ module.exports = (bookshelf, emailService, renderTemplate) => {
...
-          subject: `${appName} Invitation`,
-          html: emailBody,
-        });
+        }).then(emailBody => {
+          return emailService.send({
+            to: user.get('email'),
+            subject: `${appName} Invitation`,
+            html: emailBody,
+          }).then(data => {
+            return data;
```
Comment (spruce-bruce) [inline api/src/application/user/user-service.js]: this `.then()` call doesn't do anything

File: api/src/application/user/user-service.js [hunk 5]
```diff
@@ -30,49 +31,73 @@ module.exports = (bookshelf, emailService, renderTemplate) => {
...
+          }).catch(error => {
+            Boom.badImplementation('Something went wrong', error);
+          });
+        }).catch(error => {
+          Boom.badImplementation('Something went wrong', error);
+        });  
+      }).catch(error => {
+        Boom.badImplementation('Something went wrong', error);
```
Comment (spruce-bruce) [inline api/src/application/user/user-service.js]: I don't think these three `.catch()` calls are doing anything either, and I don't think you need them.

These `.catch()` calls create new Boom error objects but they don't `throw` them. I think the actual effect of these calls is that they will catch errors then create a new error object then discard that error object and fail silently.

If you just delete all these `.catch()` calls the framework will catch rejected promises for you and correctly return a `Boom.badImplementation()` object.

File: api/src/application/user/user-service.js [hunk 6]
```diff
@@ -30,49 +31,73 @@ module.exports = (bookshelf, emailService, renderTemplate) => {
...
 
-    resendUserInvite(userId, appName, hoursTillExpiration) {
+    resendUserInvite(userId, appName, clientId, redirect_uri, scope, hoursTillExpiration) {
       return bookshelf.model('user').where({ id: userId }).fetch().then(user => {
         if (!user) {
           return Boom.notFound();
         }
+        return clientService.findByRedirectUriAndClientId(payload.client_id, payload.redirect_uri).then(clients => {
+          if (clients.models.length === 0) {
+            return Boom.badData('The provided redirect uri is invalid for the given client id.');
```
Comment (spruce-bruce) [inline api/src/application/user/user-service.js]: Instead of returning the error object `throw` it.

`throw Boom.badData('The provided redirect uri is invalid for the given client id.');`

throwing from inside of a `.then()` call is the same thing as rejecting the promise.

File: api/src/application/user/user-service.js [hunk 7]
```diff
@@ -30,49 +31,73 @@ module.exports = (bookshelf, emailService, renderTemplate) => {
...
       let createdUser;
-      return self.create(
-        payload.email,
-        uuid.v4(),
-        {
-          app_metadata: payload.app_metadata || {},
-          profile : payload.profile || {}
+      return clientService.findByRedirectUriAndClientId(payload.client_id, payload.redirect_uri).then(clients => {
+        if (clients.models.length === 0) {
+          return Boom.badData('The provided redirect uri is invalid for the given client id.');
```
Comment (spruce-bruce) [inline api/src/application/user/user-service.js]: Same as above, throw this error instead of return

File: api/src/application/user/user-service.js [hunk 8]
```diff
@@ -30,49 +31,73 @@ module.exports = (bookshelf, emailService, renderTemplate) => {
...
+            payload.app_name,
+            payload.client_id,
+            payload.redirect_uri,
+            payload.scope,
+            payload.hours_till_expiration
+          );
+        }).then(() => createdUser).catch(error => Boom.badImplementation('Something went wrong', error));
+      }).catch(error => {
+        Boom.badImplementation('Something went wrong', error);
+      });
```
Comment (spruce-bruce) [inline api/src/application/user/user-service.js]: This `.then()` chain should be un-nested, like so:

``` javascript
return clientService.findByRedirectUriAndClientId()
  .then(clients => {
    // your validation logic
  })
  .then(() => self.create(
    payload.email,
    uuid.vr(),
    {
      app_metadata: payload.app_metadata || {},
      profile : payload.profile || {}
    }
  ))
  // etc
```

Also the two `.catch()` calls shouldn't be necessary

File: api/src/application/user/user-service.js [hunk 9]
```diff
@@ -102,8 +127,10 @@ module.exports = (bookshelf, emailService, renderTemplate) => {
     },
 
     findByEmailForOidc: function(email) {
-      return bookshelf.model('user').forge({ email_id: email.toLowerCase() }).fetch()
-        .then(user => user ? user.serialize({strictOidc: true}) : null);
+      return bookshelf.model('user').where({ email_id: email.toLowerCase() }).fetch()
+        .then(user => {
+          return user ? user.serialize({strictOidc: true}) : null
+        });
```
Comment (spruce-bruce) [inline api/src/application/user/user-service.js]: I've noticed you made a couple of these formatting changes turning calls like `() => thing` into `() => { return thing }`.

When you make formatting changes with no logic changes it bloats the code review and gives me extra stuff to look through.

File: api/migrations/20170509230917_add_email_id_column.js (hunk 2 above)
Comment (msinnes) [inline api/migrations/20170509230917_add_email_id_column.js]: The email id was discussed with Bob in order to fix the case sensitive issue on Tathata. I have added the migration for existing users back in this commit. Removed it because everything I was working on had users coming in through the seed, but this will still be needed. The seeds will have to be updated in Tathata and Strata.

**Comment by msinnes** [comment]
Design pass is complete and this is ready to be merged into release.
